packages/fxa-settings/src/lib/relier-factory.ts

```diff
@@ -0,0 +1,176 @@
+    }
+    if (flags.isOauth()) {
+      return new OAuthRelier(context);
+    }
+    if (flags.isSyncService() && flags.isV3DesktopContext()) {
+      return new BrowserRelier(context);
+    }
+
+    return new Relier(context);
+  }
+}
+
+
+
+/**
+ * Extrapolates flags from the state of the current context. The collective state of these flags are used by
+ * the factory to determine what underlying type of relier to create.
+ *
+ * Note: this logic was ported from fxa-content-server app-start.js.
+ */
+export class DefaultRelierFlags
+{
+  private get pathname() {
+    return this.urlContext.pathName;
+  }
+
+  constructor(
+    private urlContext:UrlContext,
+    private storageContext:StorageContext
+  ) { }
+
+  isDevicePairingAsAuthority() {
+    // this._searchParam('redirect_uri') ===
+    // Constants.DEVICE_PAIRING_AUTHORITY_REDIRECT_URI
+    return this._searchParam('redirect_uri') === DEVICE_PAIRING_AUTHORITY_REDIRECT_URI;
+  }
+
+  isDevicePairingAsSupplicant() {
+    return DEVICE_PAIRING_SUPPLICANT_PATHNAME_REGEXP.test(
+      this.pathname
+    );
+  }
+
+  isOauth() {
+    return (
+      !!(
+        this._searchParam('client_id') ||
+        // verification
+        this._isOAuthVerificationSameBrowser()
+      ) ||
+      this._isOAuthVerificationDifferentBrowser() ||
+      // any URL with 'oauth' in the path.
+      /oauth/.test(this.pathname)
+    );
+  }
+
+  isSyncService() {
+    return this._isService(SYNC_SERVICE)
+  }
+
+  isV3DesktopContext() {
+    return this._searchParam('context') === FX_DESKTOP_V3_CONTEXT
+  }
+
+  private _isOAuthVerificationSameBrowser() {
+    return this._isVerification() && this._isService(this._getSavedClientId());
+  }
+
+  private _isVerification() {
+    return (
+      this._isSignUpVerification() ||
+      this._isPasswordResetVerification() ||
+      this._isReportSignIn()
+    );
+  }
+
+  private _isPasswordResetVerification() {
+    return this._searchParam('code') && this._searchParam('token');
+  }
+
+  private _isReportSignIn() {
+    return this.pathname === '/report_signin';
+  }
+
+  private _isSignUpVerification() {
+    return this._searchParam('code') && this._searchParam('uid');
+  }
+
+  private _searchParam(key:string) {
+    return this.urlContext.get(key);
+  }
+
+  private _isServiceSync() {
+    return this._isService(SYNC_SERVICE);
+  }
+
+  private _isService(compareToService:string) {
+    const service = this._searchParam('service');
+    return !!(service && compareToService && service === compareToService);
+  }
+
+  private _isOAuthVerificationDifferentBrowser() {
+    return this._isVerification() && this._isServiceOAuth();
+  }
+
+  private _isServiceOAuth() {
+    const service = this._searchParam('service');
+    // any service that is not the sync service is automatically
+    // considered an OAuth service
+    return service && !this._isServiceSync();
+  }
+
+  private _getSavedClientId() {
+    return this.storageContext.get('oauth')?.client_id;
```

Comment:
I'd like to understand better why we need to do this. It required pulling the storageContext. It'd be great if the code wasn't as concerned with this. I am sure it's needed for several things though. Perhaps, the model binding that was created for reliers could also be created for important session context state.

Reply:
Are you wondering why we need to verify that browser is the same or why we're using it to verify (maybe there's a better way)?

Reply:
I'm honestly not sure why this is? Possibly it was written to storage context so that it could continue the flow if continued in another tab/window. This might be one of these things we experiment with removing and see what breaks 😅

Reply:
@LZoog  I was mostly curious why we need to verify the browser is the same. I tried various user flows locally and never saw this state in my local storage. There's probably some use case that sets it, but I don't know what that is. This is used when we call isOauth(), and some of this logic in that function seems to overlap a bit, so maybe  this check isn't super necessary 🤷? (I bet it is though.)

@vbudhram Experimentation sounds good! Do you think maybe it has something to do with sync? If any experiments come to mind that'd trigger this state in storage, let me know.